### ISSUE TITLE
PyTorch multi threading/processing behavior

### ISSUE THREAD

**Issue by dlibenzi** [pytorch/pytorch#18183]
Hi!
We are investigating PT threading performance and we are noticing a strange behavior in PT.
We have this simple benchmark, which runs in threads/processes, either a PT training loop, or a Python CPU hog.

```
import argparse

import multiprocessing as mp
import sys
import threading
import time
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision


def model_loop_fn(index, args):
  print('Thread {} started'.format(index))
  start_time = time.time()
  data = torch.randn(args.batch_size, 3, 224, 224)
  target = torch.zeros(args.batch_size, dtype=torch.int64)
  model = torchvision.models.resnet18()
  loss_fn = nn.NLLLoss()
  optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)
  for i in range(0, args.loops):
    optimizer.zero_grad()
    output = model(data)
    loss = loss_fn(output, target)
    loss.backward()
    optimizer.step()
  print('Thread {} took {:.3f}ms'.format(index,
                                         1000.0 * (time.time() - start_time)))


def hog_loop_fn(index, args):
  print('Thread {} started'.format(index))
  start_time = time.time()
  for i in range(0, args.loops):
    x = args.hog_count
    while x > 0:
      x -= 1
  print('Thread {} took {:.3f}ms'.format(index,
                                         1000.0 * (time.time() - start_time)))


def bench_resnet(args):
  loop_fn = hog_loop_fn if args.hog else model_loop_fn
  threads = []
  for i in range(0, args.num_threads):
    if args.mp:
      thread = mp.Process(target=loop_fn, args=(i, args,))
    else:
      thread = threading.Thread(target=loop_fn, args=(i, args,))
    thread.daemon = True
    threads.append(thread)
    thread.start()
  for thread in threads:
    thread.join()


if __name__ == '__main__':
  parser = argparse.ArgumentParser(add_help=False)
  parser.add_argument('--batch_size', type=int, default=4)
  parser.add_argument('--loops', type=int, default=10)
  parser.add_argument('--num_threads', type=int, default=1)
  parser.add_argument('--mp', action='store_true')
  parser.add_argument('--hog', action='store_true')
  parser.add_argument('--hog_count', type=int, default=10000000)
  args = parser.parse_args()

  bench_resnet(args)
```

These are a few runs on my workstation (12 cores - Intel(R) Xeon(R) W-2135 CPU @ 3.70GHz).
For a single thread/process HOG, times are pretty much matching, as expected:
```
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=1 --loops=10  --hog
Thread 0 started
Thread 0 took 3136.162ms

real	0m3.665s
user	0m3.588s
sys	0m0.056s
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=1 --loops=10 --mp --hog
Thread 0 started
Thread 0 took 2882.704ms

real	0m3.413s
user	0m3.314s
sys	0m0.083s
```

And the behavior changes from multi-thread to multi-process, is the expected one as well (with multi-process scaling around the GIL):
```
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=4 --loops=10  --hog
Thread 0 started
Thread 1 started
Thread 2 started
Thread 3 started
Thread 3 took 23980.427ms
Thread 1 took 24645.420ms
Thread 2 took 24684.608ms
Thread 0 took 24779.666ms

real	0m25.299s
user	0m25.153s
sys	0m0.451s
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=4 --loops=10 --mp --hog
Thread 0 started
Thread 1 started
Thread 2 started
Thread 3 started
Thread 1 took 2884.424ms
Thread 3 took 2897.024ms
Thread 0 took 2906.474ms
Thread 2 took 2909.131ms

real	0m3.431s
user	0m12.061s
sys	0m0.051s
```

Now let's look at the PT model (resnet18) benchmark loop.
For one thread/process, times are matching again, as expected;
```
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=1 --loops=10 
Thread 0 started
Thread 0 took 7160.030ms

real	0m7.701s
user	0m37.971s
sys	0m33.296s
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=1 --loops=10 --mp
Thread 0 started
Thread 0 took 7186.889ms

real	0m7.698s
user	0m53.870s
sys	0m16.627s
```

But when using 4 thread/processes, things gets interesting, as multiprocessing does not seem to be able to scale around the GIL:
```
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=4 --loops=10 
Thread 0 started
Thread 1 started
Thread 2 started
Thread 3 started
Thread 3 took 24762.609ms
Thread 0 took 24817.021ms
Thread 2 took 24832.065ms
Thread 1 took 24858.423ms

real	0m25.485s
user	2m11.353s
sys	1m38.010s
(pytorch) $ time python ~/tmp/scratch/bench_mt_resnet.py --num_threads=4 --loops=10 --mp
Thread 0 started
Thread 1 started
Thread 2 started
Thread 3 started
Thread 0 took 25734.327ms
Thread 2 took 25807.819ms
Thread 1 took 25983.238ms
Thread 3 took 25993.445ms

real	0m26.527s
user	3m53.611s
sys	1m12.719s
```

This is strange because, in case of multiprocessing, you can actually see N different python processes, which the Linux scheduler migrates to different cores.
Nevertheless the time accumulates instead of flatten (or almost - like in the HOG case).
We have seen that many (all?) code paths leaving Python and entering the ATEN API, release the GIL.
So this should in theory help with multi-threading (but does not seem so).
But the most puzzling bit is the fact that scaling does not happen even when using multi-processing.
Seems like some sort of an inter-process lock exists in PT, which prevents it from scaling in multi-processing mode.

Wondering if this behavior is known, and if you have tips to look into, in order to cut the investigation time.

**Comment by ailzhang** [comment]
cc: @apaszke @ezyang @zdevito for more ideas :)

**Comment by ezyang** [comment]
This really smells like you're bottlenecked on dataloading. What if you cut out the forward/backward loop entirely and just load samples?

**Comment by dlibenzi** [comment]
Dataloading is not existent AFAICT, as we use the same tensor within the loop.
Part of it, is PT scaling horizontally even with a single thread. Which partially explains the results.
In my workstation ~500% of CPU (out of 1200% ) is used in one thread case.
It should have headroom to scale further, but at partially explains the results.

**Comment by colesbury** [comment]
A single Python thread in PyTorch can still use multiple cores. PyTorch (and the underlying math libraries) uses OpenMP to parallelize may operations. So parallelizing on top of that can lead to oversubscription. Even in cases where you don't have oversubscription you may get diminishing returns due to shared resources (shared L3 cache, DRAM memory bandwidth).

You can limit the number of OpenMP threads using the environment variable `OMP_NUM_THREADS`. Try setting `OMP_NUM_THREADS=1` or another small number.

**Comment by dlibenzi** [comment]
A release build pushed single thread CPU utilization to ~900% in my workstation, which is closer to full saturation (1200% - 12 cores).
At this point the results can be explained to some length, so closing this one.
